skiko/src/jvmMain/cpp/windows/directXRedrawer.cc

```diff
@@ -0,0 +1,200 @@
+#ifdef SK_DIRECT3D
+#include <windows.h>
+#include <jawt_md.h>
+#include "jni_helpers.h"
+
+#include "GrBackendSurface.h"
+#include "GrDirectContext.h"
+#include "SkSurface.h"
+
+#include "d3d/GrD3DTypes.h"
+#include <d3d12sdklayers.h>
+#include "d3d/GrD3DBackendContext.h"
+#include <d3d12.h>
+#include <dxgi1_4.h>
+
+
+#define GR_D3D_CALL_ERRCHECK(X)                                        \
+    do                                                                 \
+    {                                                                  \
+        HRESULT result = X;                                            \
+        SkASSERT(SUCCEEDED(result));                                   \
+        if (!SUCCEEDED(result))                                        \
+        {                                                              \
+            SkDebugf("Failed Direct3D call. Error: 0x%08x\n", result); \
+        }                                                              \
+    } while (false)
+
+const int BuffersCount = 2;
+
+class DirectXDevice
+{
+public:
+    GrD3DBackendContext backendContext;
+    gr_cp<ID3D12Device> device;
+    gr_cp<IDXGISwapChain3> swapChain;
+    gr_cp<ID3D12CommandQueue> queue;
+    gr_cp<ID3D12Fence> fence;
+    HANDLE fenceEvent;
+    uint64_t fenceValue;
+    unsigned int bufferIndex = 1;
+    unsigned int bufferWidth = 0;
+    unsigned int bufferHeight = 0;
+
+    bool isSizeEqualTo(unsigned int width, unsigned int height)
+    {
+        return width == bufferWidth && height == bufferHeight;
+    }
+};
+
+extern "C"
+{
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_makeDirectXContext(
+        JNIEnv* env, jobject redrawer, jlong devicePtr)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+        GrD3DBackendContext backendContext = d3dDevice->backendContext;
+        return toJavaPointer(GrDirectContext::MakeDirect3D(backendContext).release());
+    }
+
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_makeDirectXRenderTarget(
+        JNIEnv * env, jobject redrawer, jlong devicePtr, jint width, jint height)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+        d3dDevice->bufferIndex = d3dDevice->swapChain->GetCurrentBackBufferIndex();
+        ID3D12Resource* buffer;
+        GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->GetBuffer(d3dDevice->bufferIndex, IID_PPV_ARGS(&buffer)));
+        GrD3DTextureResourceInfo info(buffer,
+                                      nullptr,
+                                      D3D12_RESOURCE_STATE_PRESENT,
+                                      DXGI_FORMAT_R8G8B8A8_UNORM,
+                                      1,
+                                      1,
+                                      0);
+        GrBackendRenderTarget* renderTarget = new GrBackendRenderTarget(width, height, info);
```

Comment:
We create object, but never delete it (same for DirectXDevice)

Reply:
We are deleting the render target and surface in `initCanvas`, but I noticed that we did not delete them in `SkiaLayer.dispose`, so I fixed that. Also added `DirectXDevice` removal during `Redrawer.dispose` call.